sphericart/include/cuda_cache.hpp

```diff
@@ -0,0 +1,450 @@
+
+        CUresult cuResult = driver.cuModuleLoadDataEx(&module, ptxCode.data(), 0, 0, 0);
+
+        if (cuResult != CUDA_SUCCESS) {
+            throw std::runtime_error(
+                "KernelFactory::compileAndCacheKernel: Failed to load PTX code into CUDA "
+                "module "
+                "(error code: " +
+                std::to_string(cuResult) + ")"
+            );
+        }
+
+        const char* lowered_name;
+        NVRTC_SAFE_CALL(nvrtc.nvrtcGetLoweredName(prog, this->kernel_name.c_str(), &lowered_name));
+        CUfunction kernel;
+        CUDADRIVER_SAFE_CALL(driver.cuModuleGetFunction(&kernel, module, lowered_name));
+
+        this->set(module, kernel, currentContext);
+        this->compiled = true;
+
+        NVRTC_SAFE_CALL(nvrtc.nvrtcDestroyProgram(&prog));
+    }
+
+    void initCudaDriver() {
+
+        auto& driver = CUDADriver::instance();
+
+        int deviceCount = 0;
+        // Check if CUDA has already been initialized
+        CUresult res = driver.cuDeviceGetCount(&deviceCount);
+        if (res == CUDA_ERROR_NOT_INITIALIZED) {
+            // CUDA hasn't been initialized, so we initialize it now
+            res = driver.cuInit(0);
+            if (res != CUDA_SUCCESS) {
+                throw std::runtime_error(
+                    "KernelFactory::initCudaDriver: Failed to initialize CUDA driver."
+                );
+                return;
+            }
+        }
+    }
+
+    int current_smem_size = 0;
+    int max_smem_size_optin = 0;
+    CUmodule module = nullptr;
+    CUfunction function = nullptr;
+    CUcontext context = nullptr;
+    bool compiled = false;
+
+    std::string kernel_name;
+    std::string kernel_code;
+    std::string source_name;
+    std::vector<std::string> options;
+};
+
+class CudaCacheManager {
+  public:
+    static CudaCacheManager& instance() {
+        static CudaCacheManager instance;
+        return instance;
+    }
+
+    void cacheKernel(
+        const std::string& kernel_name,
+        const std::string& source_path,
+        const std::string& source_name,
+        const std::vector<std::string>& options
+    ) {
+        kernel_cache[kernel_name] =
+            std::make_unique<CachedKernel>(kernel_name, source_path, source_name, options);
+    }
+
+    bool hasKernel(const std::string& kernel_name) const {
+        return kernel_cache.find(kernel_name) != kernel_cache.end();
+    }
+
+    CachedKernel* getKernel(const std::string& kernel_name) const {
+        auto it = kernel_cache.find(kernel_name);
+        if (it != kernel_cache.end()) {
+            return it->second.get();
+        }
+        throw std::runtime_error("Kernel not found in cache.");
+    }
+
+  private:
+    CudaCacheManager() {}
+    CudaCacheManager(const CudaCacheManager&) = delete;
+    CudaCacheManager& operator=(const CudaCacheManager&) = delete;
+
+    std::unordered_map<std::string, std::unique_ptr<CachedKernel>> kernel_cache;
+};
+
+class KernelFactory {
```

Comment:
I don't understand the difference between this and `CudaCacheManager`

Reply:
I've merged this class into KernelFactory.